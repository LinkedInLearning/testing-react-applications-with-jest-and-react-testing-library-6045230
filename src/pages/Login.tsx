import React, { useState } from 'react';
import { useNavigate, Link } from 'react-router-dom';
import styled from 'styled-components';
import { LogIn, Mail, Lock } from 'lucide-react';
import { useAuth } from '../contexts/AuthContext';

const Container = styled.div`
  max-width: 28rem;
  margin: 2rem auto;
`;

const Card = styled.div`
  background: white;
  border-radius: 1.5rem;
  box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06);
  overflow: hidden;
`;

const Header = styled.div`
  background: linear-gradient(135deg, #2563eb, #1d4ed8);
  padding: 2rem;
  text-align: center;
  color: white;
`;

const Title = styled.h1`
  font-size: 1.875rem;
  font-weight: 700;
  margin-bottom: 0.5rem;
`;

const Subtitle = styled.p`
  color: rgba(255, 255, 255, 0.9);
  font-size: 0.875rem;
`;

const Form = styled.form`
  padding: 2rem;
`;

const FormGroup = styled.div`
  margin-bottom: 1.5rem;
  position: relative;
`;

const Input = styled.input`
  width: 100%;
  padding: 0.75rem 1rem 0.75rem 2.5rem;
  border: 2px solid #e5e7eb;
  border-radius: 0.75rem;
  font-size: 1rem;
  transition: all 0.2s;

  &:focus {
    outline: none;
    border-color: #2563eb;
    box-shadow: 0 0 0 3px rgba(37, 99, 235, 0.1);
  }
`;

const InputIcon = styled.div`
  position: absolute;
  left: 0.75rem;
  top: 50%;
  transform: translateY(-50%);
  color: #6b7280;
`;

const ErrorMessage = styled.div`
  background-color: #fee2e2;
  color: #dc2626;
  padding: 1rem;
  border-radius: 0.75rem;
  margin-bottom: 1.5rem;
  font-size: 0.875rem;
  display: flex;
  align-items: center;
  gap: 0.5rem;
`;

const Button = styled.button`
  width: 100%;
  padding: 0.875rem;
  background: linear-gradient(135deg, #2563eb, #1d4ed8);
  color: white;
  border: none;
  border-radius: 0.75rem;
  font-weight: 600;
  font-size: 1rem;
  cursor: pointer;
  transition: all 0.2s;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 0.5rem;

  &:hover {
    transform: translateY(-1px);
    box-shadow: 0 4px 6px -1px rgba(37, 99, 235, 0.2);
  }

  &:disabled {
    opacity: 0.7;
    cursor: not-allowed;
    transform: none;
  }
`;

const LinkText = styled.div`
  text-align: center;
  margin-top: 1.5rem;
  color: #6b7280;
  font-size: 0.875rem;

  a {
    color: #2563eb;
    text-decoration: none;
    font-weight: 500;
    margin-left: 0.25rem;

    &:hover {
      text-decoration: underline;
    }
  }
`;

export function Login() {
  const navigate = useNavigate();
  const { signIn } = useAuth();
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState('');
  const [loading, setLoading] = useState(false);

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    try {
      setError('');
      setLoading(true);
      await signIn(email, password);
      navigate('/');
    } catch (error) {
      setError('Invalid email or password');
    } finally {
      setLoading(false);
    }
  };

  return (
    <Container>
      <Card>
        <Header>
          <Title>Welcome Back</Title>
          <Subtitle>Sign in to continue to your account</Subtitle>
        </Header>

        <Form onSubmit={handleSubmit}>
          {error && (
            <ErrorMessage>
              <LogIn size={18} />
              {error}
            </ErrorMessage>
          )}

          <FormGroup>
            <InputIcon>
              <Mail size={18} />
            </InputIcon>
            <Input
              type="email"
              placeholder="Email address"
              value={email}
              onChange={(e) => setEmail(e.target.value)}
              required
            />
          </FormGroup>

          <FormGroup>
            <InputIcon>
              <Lock size={18} />
            </InputIcon>
            <Input
              type="password"
              placeholder="Password"
              value={password}
              onChange={(e) => setPassword(e.target.value)}
              required
            />
          </FormGroup>

          <Button type="submit" disabled={loading}>
            {loading ? (
              'Signing in...'
            ) : (
              <>
                <LogIn size={18} />
                Sign In
              </>
            )}
          </Button>

          <LinkText>
            Don't have an account?
            <Link to="/register">Create account</Link>
          </LinkText>
        </Form>
      </Card>
    </Container>
  );
}