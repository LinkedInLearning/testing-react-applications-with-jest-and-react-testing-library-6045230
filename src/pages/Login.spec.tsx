import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { describe, vi, it, expect } from 'vitest';
import { Login } from './Login';
import userEvent from '@testing-library/user-event';

// Mock functions we'll configure in each test
const mockNavigate = vi.fn()
const mockSignIn = vi.fn()


// Mock the modules we need to control
vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});

vi.mock('../contexts/AuthContext', () => ({
  useAuth: () => ({
    signIn: mockSignIn,
  }),
}));


// Test constants
const validCredentials = {
  email: 'test@example.com',
  password: 'Password123'
};

// Helper to render component with necessary providers
const renderLogin = () => {
  return render(
    <MemoryRouter>
      <Login />
    </MemoryRouter>
  );
};

// Helper to fill form fields
const fillLoginForm = async (email = validCredentials.email, password = validCredentials.password) => {
  const user = userEvent.setup();
  const emailInput = screen.getByPlaceholderText('Email address');
  const passwordInput = screen.getByPlaceholderText('Password');

  await user.type(emailInput, email);
  await user.type(passwordInput, password);

  return { emailInput, passwordInput };
};

// Helper to submit the form
const submitForm = async () => {
  const user = userEvent.setup();
  const submitButton = screen.getByRole('button', { name: /sign in/i });
  await user.click(submitButton);
  return submitButton;
};

describe('Login Component Snapshots', () => {
  it('renders initial login form correctly', () => {
    const { asFragment } = renderLogin();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders error state correctly', async () => {
    // Configure the mock to throw an error
    mockSignIn.mockRejectedValueOnce(new Error('Invalid credentials'));

    const { asFragment } = renderLogin();

    // Fill and submit form to trigger error state
    await fillLoginForm();
    await submitForm();

    // Wait for error to appear
    await screen.findByText('Invalid email or password');

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders loading state correctly', async () => {
    // Configure the mock to return a pending promise that won't resolve during the test
    mockSignIn.mockImplementationOnce(() => new Promise(() => { }));

    const { asFragment } = renderLogin();

    // Fill and submit form to trigger loading state
    await fillLoginForm();
    await submitForm();

    // Verify loading state is showing
    expect(screen.getByText('Signing in...')).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Login Component Functionality', () => {
  it('submits the form with correct email and password', async () => {
    renderLogin();

    // Fill and submit the form
    await fillLoginForm();
    await submitForm();

    // Verify signIn was called with correct values
    expect(mockSignIn).toHaveBeenCalledWith(
      validCredentials.email,
      validCredentials.password
    );
  });

  it('shows error message when authentication fails', async () => {
    // Configure mock to reject
    mockSignIn.mockRejectedValueOnce(new Error('Invalid credentials'));

    renderLogin();

    // Fill and submit the form
    await fillLoginForm();
    await submitForm();

    // Verify error message appears
    const errorMessage = await screen.findByText('Invalid email or password');
    expect(errorMessage).toBeInTheDocument();
  });

  it('navigates to homepage after successful login', async () => {
    // Configure mock to resolve
    mockSignIn.mockResolvedValueOnce({});

    renderLogin();

    // Fill and submit the form
    await fillLoginForm();
    await submitForm();

    // Verify navigation occurred after successful login
    await waitFor(() => {
      expect(mockNavigate).toHaveBeenCalledWith('/');
    });
  });

  it('disables button and shows loading state during authentication', async () => {
    // Set up a promise we can control
    let resolveSignIn: (value?: unknown) => void = () => { };
    mockSignIn.mockImplementationOnce(() => new Promise(resolve => {
      resolveSignIn = resolve;
    }));

    renderLogin();

    // Fill and submit the form
    await fillLoginForm();
    const button = await submitForm();

    // Verify button is disabled and shows loading text
    expect(button).toBeDisabled();
    expect(screen.getByText('Signing in...')).toBeInTheDocument();

    // Resolve the promise to complete the authentication
    resolveSignIn({});

    // Verify loading state is removed
    await waitFor(() => {
      expect(screen.queryByText('Signing in...')).not.toBeInTheDocument();
    });
  });

  it('validates that form fields are required', () => {
    renderLogin();

    // Get inputs directly to check their attributes
    const emailInput = screen.getByPlaceholderText('Email address');
    const passwordInput = screen.getByPlaceholderText('Password');

    // Verify required attribute is present
    expect(emailInput).toHaveAttribute('required');
    expect(passwordInput).toHaveAttribute('required');
  });

  it('clears error message when form is resubmitted', async () => {
    // First submission fails
    mockSignIn.mockRejectedValueOnce(new Error('Invalid credentials'));

    renderLogin();

    // Fill and submit to trigger error
    await fillLoginForm();
    await submitForm();

    // Verify error appears
    const errorMessage = await screen.findByText('Invalid email or password');
    expect(errorMessage).toBeInTheDocument();

    // Mock second submission to resolve
    mockSignIn.mockResolvedValueOnce({});

    // Submit again
    submitForm();

    // Verify error is cleared
    await waitFor(() => {
      expect(screen.queryByText('Invalid email or password')).not.toBeInTheDocument();
    });
  });
});