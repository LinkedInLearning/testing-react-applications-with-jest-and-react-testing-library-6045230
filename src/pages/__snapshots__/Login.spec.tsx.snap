// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login Component Snapshots > renders error state correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-VHjGu dUAksA"
  >
    <div
      class="sc-gnOvAp jrVZLq"
    >
      <div
        class="sc-dQkurY ckYwKI"
      >
        <h1
          class="sc-glJfXD keBBgA"
        >
          Welcome Back
        </h1>
        <p
          class="sc-dOfePm enCxEj"
        >
          Sign in to continue to your account
        </p>
      </div>
      <form
        class="sc-jgmqJh bFKKOZ"
      >
        <div
          class="sc-jRLeyu eKKQYG"
        >
          <svg
            class="lucide lucide-log-in "
            fill="none"
            height="18"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 3h4a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2h-4"
            />
            <polyline
              points="10 17 15 12 10 7"
            />
            <line
              x1="15"
              x2="3"
              y1="12"
              y2="12"
            />
          </svg>
          Invalid email or password
        </div>
        <div
          class="sc-gIIpAQ kQrbOK"
        >
          <div
            class="sc-ihGilW eqjTkv"
          >
            <svg
              class="lucide lucide-mail "
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="16"
                rx="2"
                width="20"
                x="2"
                y="4"
              />
              <path
                d="m22 7-8.97 5.7a1.94 1.94 0 0 1-2.06 0L2 7"
              />
            </svg>
          </div>
          <input
            class="sc-kFkjun sNfA-d"
            placeholder="Email address"
            required=""
            type="email"
            value="test@example.com"
          />
        </div>
        <div
          class="sc-gIIpAQ kQrbOK"
        >
          <div
            class="sc-ihGilW eqjTkv"
          >
            <svg
              class="lucide lucide-lock "
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="11"
                rx="2"
                ry="2"
                width="18"
                x="3"
                y="11"
              />
              <path
                d="M7 11V7a5 5 0 0 1 10 0v4"
              />
            </svg>
          </div>
          <input
            class="sc-kFkjun sNfA-d"
            placeholder="Password"
            required=""
            type="password"
            value="Password123"
          />
        </div>
        <button
          class="sc-cRHGzb STQdD"
          type="submit"
        >
          <svg
            class="lucide lucide-log-in "
            fill="none"
            height="18"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 3h4a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2h-4"
            />
            <polyline
              points="10 17 15 12 10 7"
            />
            <line
              x1="15"
              x2="3"
              y1="12"
              y2="12"
            />
          </svg>
          Sign In
        </button>
        <div
          class="sc-jOTjWM hPbKnK"
        >
          Don't have an account?
          <a
            href="/register"
          >
            Create account
          </a>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Login Component Snapshots > renders initial login form correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-VHjGu dUAksA"
  >
    <div
      class="sc-gnOvAp jrVZLq"
    >
      <div
        class="sc-dQkurY ckYwKI"
      >
        <h1
          class="sc-glJfXD keBBgA"
        >
          Welcome Back
        </h1>
        <p
          class="sc-dOfePm enCxEj"
        >
          Sign in to continue to your account
        </p>
      </div>
      <form
        class="sc-jgmqJh bFKKOZ"
      >
        <div
          class="sc-gIIpAQ kQrbOK"
        >
          <div
            class="sc-ihGilW eqjTkv"
          >
            <svg
              class="lucide lucide-mail "
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="16"
                rx="2"
                width="20"
                x="2"
                y="4"
              />
              <path
                d="m22 7-8.97 5.7a1.94 1.94 0 0 1-2.06 0L2 7"
              />
            </svg>
          </div>
          <input
            class="sc-kFkjun sNfA-d"
            placeholder="Email address"
            required=""
            type="email"
            value=""
          />
        </div>
        <div
          class="sc-gIIpAQ kQrbOK"
        >
          <div
            class="sc-ihGilW eqjTkv"
          >
            <svg
              class="lucide lucide-lock "
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="11"
                rx="2"
                ry="2"
                width="18"
                x="3"
                y="11"
              />
              <path
                d="M7 11V7a5 5 0 0 1 10 0v4"
              />
            </svg>
          </div>
          <input
            class="sc-kFkjun sNfA-d"
            placeholder="Password"
            required=""
            type="password"
            value=""
          />
        </div>
        <button
          class="sc-cRHGzb STQdD"
          type="submit"
        >
          <svg
            class="lucide lucide-log-in "
            fill="none"
            height="18"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 3h4a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2h-4"
            />
            <polyline
              points="10 17 15 12 10 7"
            />
            <line
              x1="15"
              x2="3"
              y1="12"
              y2="12"
            />
          </svg>
          Sign In
        </button>
        <div
          class="sc-jOTjWM hPbKnK"
        >
          Don't have an account?
          <a
            href="/register"
          >
            Create account
          </a>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Login Component Snapshots > renders loading state correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-VHjGu dUAksA"
  >
    <div
      class="sc-gnOvAp jrVZLq"
    >
      <div
        class="sc-dQkurY ckYwKI"
      >
        <h1
          class="sc-glJfXD keBBgA"
        >
          Welcome Back
        </h1>
        <p
          class="sc-dOfePm enCxEj"
        >
          Sign in to continue to your account
        </p>
      </div>
      <form
        class="sc-jgmqJh bFKKOZ"
      >
        <div
          class="sc-gIIpAQ kQrbOK"
        >
          <div
            class="sc-ihGilW eqjTkv"
          >
            <svg
              class="lucide lucide-mail "
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="16"
                rx="2"
                width="20"
                x="2"
                y="4"
              />
              <path
                d="m22 7-8.97 5.7a1.94 1.94 0 0 1-2.06 0L2 7"
              />
            </svg>
          </div>
          <input
            class="sc-kFkjun sNfA-d"
            placeholder="Email address"
            required=""
            type="email"
            value="test@example.com"
          />
        </div>
        <div
          class="sc-gIIpAQ kQrbOK"
        >
          <div
            class="sc-ihGilW eqjTkv"
          >
            <svg
              class="lucide lucide-lock "
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="11"
                rx="2"
                ry="2"
                width="18"
                x="3"
                y="11"
              />
              <path
                d="M7 11V7a5 5 0 0 1 10 0v4"
              />
            </svg>
          </div>
          <input
            class="sc-kFkjun sNfA-d"
            placeholder="Password"
            required=""
            type="password"
            value="Password123"
          />
        </div>
        <button
          class="sc-cRHGzb STQdD"
          disabled=""
          type="submit"
        >
          Signing in...
        </button>
        <div
          class="sc-jOTjWM hPbKnK"
        >
          Don't have an account?
          <a
            href="/register"
          >
            Create account
          </a>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
